### ISSUE TITLE
Boolean mask doe snot work

### ISSUE THREAD

**Issue by jfpuget** [lucidrains/nGPT-pytorch#11]
on A100, using nvidia/pytroch 24.09.py3 container, pasisng a Bollean mask does not work with sdpa. It yields NaN gradients.

I made it work by multiplying the mask by -10. (-20. works too).

I tested the fix: 

`mask = mask.float() * -torch.finfo(q.dtype).max`

And this also yields NaN gradients.

What works for me is simply:

`mask = mask * -10.0`

**Comment by lucidrains** [comment]
i see, `-10` may not be sufficiently large

**Comment by lucidrains** [comment]
@jfpuget how about `-1e10` ?

**Comment by lucidrains** [comment]
@jfpuget are you sure the entire row is not masked out by accident? run a `assert mask.any(dim = -1).all()`

**Comment by jfpuget** [comment]
The mask is added to attention score before softmax, i.e. before exponentiation. e**-10 is already quite small. 

-100 works too.

-1e-10 yields Nan gradients.

**Comment by lucidrains** [comment]
@jfpuget yes, but the query key similarity score (if not qk normed) could have a wide range

**Comment by lucidrains** [comment]
@jfpuget definitely double check to make sure the row isn't completely masked out

**Comment by jfpuget** [comment]
You were right. I use a 2D mask, and in some cases a row or column is fully masked. But in that case the loss is also masked for the same row or column. I don't get why there was an issue. 

anyway, I found a way to avoid fully masked rows or columns, and no longer have NaN when using a Boolean mask. Fp16 works fine too.

I'll double check tomorrow.

**Comment by lucidrains** [comment]
@jfpuget thank you! I can add a guard for that and just zero out the all masked out rows

I think future versions of pytorch may have a flag to guard against this edge case too

**Comment by jfpuget** [comment]
I still don't get why attention in a fully masked row impacts attention in other rows. I'll dig.

**Comment by lucidrains** [comment]
should be resolved with an extra guard in the latest!
